cmd/fed_serve.go

```diff
@@ -0,0 +1,213 @@
+//go:build !windows
+
+/***************************************************************
+ *
+ * Copyright (C) 2023, Pelican Project, Morgridge Institute for Research
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you
+ * may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ***************************************************************/
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/pelicanplatform/pelican/config"
+	"github.com/pelicanplatform/pelican/param"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+)
+
+var currentModules = map[string]string{"registry": "", "director": "", "origin": ""}
+var cModsStringList = "registry, director, origin"
+
+func fedCheckWorking(reqUrl string, server string) error {
+	expiry := time.Now().Add(10 * time.Second)
+	success := false
+	for !(success || time.Now().After(expiry)) {
+		time.Sleep(50 * time.Millisecond)
+		req, err := http.NewRequest("GET", reqUrl, nil)
+		if err != nil {
+			return err
+		}
+		httpClient := http.Client{
+			Transport: config.GetTransport(),
+			Timeout:   50 * time.Millisecond,
+		}
+		_, err = httpClient.Do(req)
+		if err != nil {
+			log.Infoln("Failed to send request to "+server+"; likely server is not up (will retry in 50ms):", err)
+		} else {
+			success = true
+			log.Debugln(server + " server appears to be functioning")
+		}
+	}
+
+	return nil
+}
+
+func fedServeStart(cmd *cobra.Command, args []string) error {
+	moduleSlice, err := cmd.Flags().GetStringSlice("modules")
+	if err != nil {
+		return errors.Wrap(err, "Failed to load modules passed via --module flag")
+	}
+	// Faking a set of certain values
+	var moduleMap map[string]uint16 = map[string]uint16{}
```

Comment:
Let's convert this to an internal struct.  Something along the lines of:

```
type modulePorts struct {
  RegistryPort uint16
  DirectorPort uint16
  ...
}
```

This introduces some type safety and, I suspect, will result in a cleaner syntax.